### PR TITLE
BlockBuilder: Uploads of blocks once they are compacted

### PR DESCRIPTION
**Pull request by codesome**
@narqo 

We follow a simple approach: upload the blocks as soon as they are created. Since we need the blocks to be uploaded before we do a kafka commit, there is no point of using the shipper code.

### CODE REVIEW

File: pkg/storage/shipper/shipper.go
Comment (narqo) [inline pkg/storage/shipper/shipper.go:6]: > Extracting `shipper` from ingester to a dedicated package

If we want to go this route, I suggest we do that separately, landing this to the `main` branch first. Otherwise, it will be fairy hard to merge `blockbuilder` back to `main` later

**Comment by codesome** [comment]
Converted to Draft since now the plan is to upload the blocks in sync with compactions

**Comment by codesome** [comment]
@narqo should be ready now. Much simpler.

File: pkg/blockbuilder/tsdb.go
Comment (codesome) [inline pkg/blockbuilder/tsdb.go:263]: Only reason to take a `blockUploaderForUser` here instead of simply uploading is to make writing tests easier with a mock uploader. (might need more changes in BlockBuilder for this, but we can do that when writing tests for BlockBuilder)

**Review by narqo — APPROVED** [review]
🔥 looks much simpler, indeed

**Comment by codesome** [comment]
I will merge this after merging https://github.com/grafana/mimir/pull/8296. Fixing conflicts in this PR will be easier than that one.

**Pull request merged**
